Alirezazahiri/feat/add auth#2

alirezazahiri opened the pull request:
- Added database (PostgresQL) connection.
- Added JWT configuration to yaml 
- Added config loader for env variables related to JWT tokens
- Added database query preparation for better error handling
- Renamed "identifier" field to "email" for registration (the field we get from the user in request body)
- Refactored JWT to become a dependency that accepts a configuration struct (access_token_expiration_time_ns, jwt_secret, etc.)
- Added "mobile" field to user struct 
- Added postgresql migration file related to adding a new column, named "mobile" in the users table

alirezazahiri committed:
feat: implement user registration + unique username generation

alirezazahiri committed:
feat: add jwt authentication + configurable secret and expiration

alirezazahiri committed:
refactor: improve JWT configuration + database query preparation

atareversei commented on internal/repository/postgresql/postgresqluser/user.go:
```diff
@@ -0,0 +1,159 @@
+	err = row.Scan(userId)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return false, richerror.New(op).
```
This is not an error
I mean we use the `richerror` to capture server errors and this is not a server error, we simply didn't find the user with the given hash, so I think `false, nil` is a better return value

atareversei commented on internal/service/userservice/register.go:
```diff
@@ -2,11 +2,53 @@ package userservice
+		User: userentity.User{
+			Email:    req.Email,
+			UserName: username.GenerateUnique(usernameBase),
+			Mobile:   "",
```
If we are not setting it, we can remove it from here to save one line of code

atareversei commented on internal/repository/postgresql/postgresqluser/user.go:
```diff
@@ -0,0 +1,159 @@
+	return true, nil
+}
+
+func (d *DB) FindUserByMobile(ctx context.Context, mobile string) (userentity.UserWithPasswordHash, bool, error) {
```
We should return `entity.User` not `entity.UserWithHashedPassword`

atareversei commented on internal/repository/postgresql/postgresqluser/user.go:
```diff
@@ -0,0 +1,159 @@
+	}
+
+	row := stmt.QueryRowContext(ctx, normalizedMobile)
+	var user userentity.UserWithPasswordHash
+	err = row.Scan(&user.UserName, &user.Email, &user.PasswordHash)
+	user.Mobile = normalizedMobile
+
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return userentity.UserWithPasswordHash{}, false, nil
+		}
+		return userentity.UserWithPasswordHash{}, false, richerror.New(op).
+			WithKind(richerror.KindUnexpected).
+			WithMessage(errmsg.CantScanQueryResult).
+			WithError(err)
+	}
```
could we turn this into a reusable function?

atareversei commented on internal/repository/postgresql/postgresqluser/user.go:
```diff
@@ -0,0 +1,159 @@
+	return user, true, nil
+}
+
+func (d *DB) FindUserByEmail(ctx context.Context, email string) (userentity.UserWithPasswordHash, bool, error) {
```
The same problem with `enitity.UserWithHashedPassword`

atareversei commented on internal/repository/postgresql/postgresqluser/user.go:
```diff
@@ -0,0 +1,159 @@
+	row := stmt.QueryRowContext(ctx, email)
+
+	var user userentity.UserWithPasswordHash
+	err = row.Scan(&user.UserName, &user.Email, &user.PasswordHash)
```
Turning this into a function + we should scan for user's mobile as well (we are returning the user's mobile in the method above)

atareversei commented on internal/repository/postgresql/postgresqluser/user.go:
```diff
@@ -0,0 +1,159 @@
+	return user, true, nil
+}
+
+func (d *DB) FindUserByUserName(ctx context.Context, username string) (userentity.UserWithPasswordHash, bool, error) {
```
The same problem with the return type and scan redundancy

atareversei commented on main.go:
```diff
@@ -31,8 +32,9 @@ func main() {
+	jwt := jwtutil.New(cfg.Auth.JWT)
 
-	userSvc := userservice.New(userRepo)
+	userSvc := userservice.New(userRepo, jwt)
```
Why do we need jwt as a dependency? And if we need it as a dependency, shouldn't we turn it into an interface?

alirezazahiri replied:
One other solution that I have in mind is "making it a globally accessible singleton, like we did with translation package".

Let me know if you find this solution more intriguing!

alirezazahiri committed:
refactor: enhance user authentication and repository methods

- Modify user repository methods to support flexible identifier-based authentication
- Update login service to use new repository methods with dynamic identifier types
- Simplify user entity and query scanning logic
- Improve password verification and login flow

alirezazahiri committed:
refactor: simplify JWT initialization and usage

- Move JWT configuration to a global singleton approach
- Remove JWT dependency from user service constructor
- Update JWT generation and decoding methods to use global configuration
- Modify main initialization to use new JWT setup method
